minikube-gradle-plugin/src/main/java/com/google/cloud/tools/minikube/util/CommandExecutor.java

```diff
@@ -0,0 +1,163 @@
+public class CommandExecutor {
+  public CommandExecutor() {}
+
+  public CommandExecutor setLogger(Logger logger) {
+    this.logger = logger;
+    return this;
+  }
+
+  @VisibleForTesting static final int TIMEOUT_SECONDS = 5;
+
+  @VisibleForTesting
+  static class ProcessBuilderFactory {
+    ProcessBuilder createProcessBuilder() {
+      return new ProcessBuilder();
+    }
+  }
+
+  @VisibleForTesting
+  CommandExecutor setProcessBuilderFactory(ProcessBuilderFactory processBuilderFactory) {
+    this.processBuilderFactory = processBuilderFactory;
+    return this;
+  }
+
+  @VisibleForTesting
+  static class ExecutorServiceFactory {
+    ExecutorService createExecutorService() {
+      return Executors.newSingleThreadExecutor();
+    }
+  }
+
+  @VisibleForTesting
+  CommandExecutor setExecutorServiceFactory(ExecutorServiceFactory executorServiceFactory) {
+    this.executorServiceFactory = executorServiceFactory;
+    return this;
+  }
+
+  private ProcessBuilderFactory processBuilderFactory = new ProcessBuilderFactory();
+  private ExecutorServiceFactory executorServiceFactory = new ExecutorServiceFactory();
+  private Logger logger;
+
+  /**
+   * Runs the command.
+   *
+   * @param command the list of command line tokens
+   * @return the output of the command as a list of lines
+   * @throws GradleException if the command exited with non-zero exit code
+   */
+  public List<String> run(List<String> command) throws IOException, InterruptedException {
+    if (logger != null) {
+      logger.debug("Running command : " + String.join(" ", command));
+    }
+
+    ExecutorService executor = executorServiceFactory.createExecutorService();
+
+    // Builds the command to execute.
+    final Process process = buildProcess(command);
+
+    // Runs the command and streams the output.
+    List<String> output = new ArrayList<>();
+    int exitCode = runProcessWithExecutor(process, executor, output);
+    shutdownExecutor(executor);
+
+    // Stops the build if the command fails to do something, we may want to make this configurable.
+    if (exitCode != 0) {
+      throw new GradleException("command exited with non-zero exit code : " + exitCode);
+    }
+
+    return output;
+  }
+
+  private Process buildProcess(List<String> command) throws IOException {
+    ProcessBuilder processBuilder = processBuilderFactory.createProcessBuilder();
+    processBuilder.command(command);
+    processBuilder.redirectErrorStream(true);
+    return processBuilder.start();
+  }
+
+  /**
+   * Helper function that runs a process on an {@code ExecutorService} and stores the output as a
+   * list of lines
+   *
+   * @param process the process to run
+   * @param executor the {@code ExecutorService} that executes the process
+   * @param output the {@code List} to store the output in
+   * @return the exit code of the process
+   */
+  private int runProcessWithExecutor(Process process, ExecutorService executor, List<String> output)
```

Comment:
I don't think you need to separate this function out, it seems like it can just sit in `run` with the rest of the stuff.